### ISSUE TITLE
Unpaired Bidi Control Character Detection

### ISSUE THREAD

**Issue by GBeauregard** [PyCQA/pylint#5281]
### Current problem

I'm trying to detect if my source code contains unpaired bidi control characters a la https://trojansource.codes/

### Desired solution

Although some languages (rust) and compilers (clang, gcc) are being modified to reject unpaired bidi characters, the python security team rejected this and the discussion is that this should be implemented in python linters instead: https://mail.python.org/archives/list/python-dev@python.org/thread/6DBJJRQHA2SP5Q27MOMDSTCOXMW7ITNR/

Therefore, I'd like pylint to implement a check for this.

**Comment by DudeNr33** [comment]
I think https://github.com/PyCQA/bandit would probably more suitable for this, at it focuses on security issues in Python Code.

**Comment by GBeauregard** [comment]
> I think https://github.com/PyCQA/bandit would probably be more suitable for this, at it focuses on security issues in Python Code.

Bandit claims to work on the AST which I don't think is helpful here, but maybe they can expand? Ideally we paint a wide brush on the python ecosystem with this check if the python compiler isn't rejecting it.

**Comment by Pierre-Sassoulas** [comment]
Could you provide some example of such unpaired bidi control characters targeting python ? The C comment trick from the article you gave would not work for example. This concept is a little fuzzy for me, and example would come handy to create functional tests and would help implementation too.

**Comment by cdce8p** [comment]
They uploaded some examples to their Github repo here: https://github.com/nickboucher/trojan-source/tree/main/Python

After reading an article, I also though that this might be something a linter could detect. Not sure though if `pylint` is the correct one for it. Might be better to write a completely new one if nobody has done so already.

**Comment by Pierre-Sassoulas** [comment]
> They uploaded some examples to their Github repo here: https://github.com/nickboucher/trojan-source/tree/main/Python

Github detects two of the examples as misleading !

**Comment by GBeauregard** [comment]
> > They uploaded some examples to their Github repo here: https://github.com/nickboucher/trojan-source/tree/main/Python
> 
> Github detects two of the examples as misleading !

Yes, these checks were added a few days ago as a result of the linked work.

**Comment by GBeauregard** [comment]
> After reading an article, I also though that this might be something a linter could detect. Not sure though if `pylint` is the correct one for it. Might be better to write a completely new one if nobody has done so already.

My main concern is I'd like this fixed for the Python ecosystem and not just my personal projects/the few that go find security linters. It's a bit unfortunate CPython punted this to linters when as far as I know it's not possible to generate any well formed false positives (valid bidi usage is paired).

**Comment by CarliJoy** [comment]
I created a plugin that checks for bidi controls.
As far I understood also correct bidi controls can be an issues, as they could lead to unexpected displays.
See https://www.python.org/dev/peps/pep-0672/ for details.

Besides this, the complete issues is also about non readable code, i.e. if you have homeglyph characters.
For this I also wrote an plugin that enforces ASCII only chars (in case you program in english only).

Also I create a plugin that checks for invalid ASCII chars, mention in PEP 672.

All plugins have proper tests of course.

I could create a merge request to include these plugins as checkers into pylint. @Pierre-Sassoulas should I do this?
Some of the test could/should be optional, we could discuss in the MR how to continue with them.
But I won't do the effort of the MR if you think that there is no place for it anyway.
